File: src/lib/dependency-resolver.ts
/**
 * Frontend dependency resolver for mods (TASK-22).
 * Uses Modrinth API to fetch dependency chains and Rust backend to check
 * which deps are already installed.
 */

import { invoke } from "@tauri-apps/api/core";
import { getProjectVersions } from "@/lib/modrinth";
import type { ModrinthVersion } from "@/types/modrinth";

// ─── Simple check (used by install flow) ───────────────────────────

/**
 * Simple check: given a dependencies array from a Modrinth version,
 * check which are installed via the Rust backend.
 * Returns the same structure as the recursive resolve but without transitive resolution.
 */
export async function checkModDependencies(
  instanceName: string,
  dependencies: ModrinthVersion["dependencies"],
): Promise<ResolveResult> {
  const relevantDeps = dependencies.filter(
    (d) => d.project_id && d.dependency_type !== "embedded",
  );

  if (relevantDeps.length === 0) {
    return { dependencies: [], circularDetected: false, hasMissing: false, hasConflicts: false };
  }

  // Resolve project names AND slugs (slugs needed for Rust matching)
  const projectIds = [...new Set(relevantDeps.map((d) => d.project_id as string))];
  const metaResults = await Promise.all(projectIds.map((id) => resolveProjectMeta(id)));

  // Maps keyed by hash project_id
  const slugMap = new Map<string, string | null>();
  const metaMap = new Map<string, { name: string; iconUrl: string | null }>();
  for (let i = 0; i < projectIds.length; i++) {
    const r = metaResults[i];
    if (r) {
      slugMap.set(projectIds[i]!, r.slug);
      metaMap.set(projectIds[i]!, { name: r.name, iconUrl: r.iconUrl });
    }
  }

  // Send SLUGS to Rust (Rust matches against registry's project_slug)
  const statusMap = await checkInstalledStatus(
    instanceName,
    relevantDeps.map((d) => {
      const hash = d.project_id!;
      const slug = slugMap.get(hash);
      return { projectId: slug ?? hash, type: d.dependency_type };
    }),
  );

  // Build deps — look up status by slug, but keep project_id (hash) for identity
  const deps: DependencyInfo[] = relevantDeps.map((dep) => {
    const pid = dep.project_id!;
    const slug = slugMap.get(pid);
    // Try to look up by slug first, then by hash (fallback)
    const status = statusMap.get(slug ?? pid) ?? statusMap.get(pid);
    const meta = metaMap.get(pid);
    return {
      projectId: pid,
      type: dep.dependency_type as DependencyInfo["type"],
      installed: status?.installed ?? false,
      installedName: status?.installedName,
      modName: meta?.name ?? pid,
      iconUrl: meta?.iconUrl ?? null,
      depth: 0,
    };
  });

  const hasMissing = deps.some((d) => d.type === "required" && !d.installed);
  const hasConflicts = deps.some((d) => d.type === "incompatible" && d.installed);

  return { dependencies: deps, circularDetected: false, hasMissing, hasConflicts };
}

// ─── Types ──────────────────────────────────────────────────────────

export interface DependencyInfo {
  /** Modrinth project ID */
  projectId: string;
  /** Dependency type */
  type: "required" | "optional" | "incompatible" | "embedded";
  /** Whether already installed in instance */
  installed: boolean;
  /** Installed mod name if installed */
  installedName?: string;
  /** Resolved mod name (from API) */
  modName?: string;
  /** Resolved mod icon URL */
  iconUrl?: string | null;
  /** Child dependencies (transitive) */
  children?: DependencyInfo[];
  /** Depth level for UI indentation */
  depth: number;
}

export interface ResolveResult {
  dependencies: DependencyInfo[];
  circularDetected: boolean;
  hasMissing: boolean;
  hasConflicts: boolean;
}

// ─── Helpers ────────────────────────────────────────────────────────

/** Match a project slug to installed mods by comparing against installed projectSlugs. */
async function checkInstalledStatus(
  instanceName: string,
  deps: Array<{ projectId: string; type: string }>,
): Promise<Map<string, { installed: boolean; installedName?: string }>> {
  try {
    const result = await invoke<{
      dependencies: Array<{
        projectId: string;
        dependencyType: string;
        installed: boolean;
        installedName?: string | null;
      }>;
    }>("resolve_mod_dependencies", {
      instanceName,
      dependencies: deps.map((d) => ({
        projectId: d.projectId,
        dependencyType: d.type,
      })),
    });

    const map = new Map<string, { installed: boolean; installedName?: string }>();
    for (const dep of result.dependencies) {
      map.set(dep.projectId, {
        installed: dep.installed,
        installedName: dep.installedName ?? undefined,
      });
    }
    return map;
  } catch {
    return new Map();
  }
}

/** Resolve a single project to get its metadata (name, icon, slug).
 *  Slug is critical: Rust backend matches installed mods by project_slug,
 *  but Modrinth dependency API returns project_id (hash). We resolve it here. */
async function resolveProjectMeta(
  projectId: string,
): Promise<{ name: string; iconUrl: string | null; slug: string | null } | null> {
  try {
    const response = await fetch(
      `https://api.modrinth.com/v2/project/${encodeURIComponent(projectId)}`,
      { signal: AbortSignal.timeout(5000) },
    );
    if (!response.ok) return null;
    const data = await response.json();
    return {
      name: data.title || projectId,
      iconUrl: data.icon_url || null,
      slug: data.slug || null,
    };
  } catch {
    return { name: projectId, iconUrl: null, slug: null };
  }
}

// ─── Main resolver ──────────────────────────────────────────────────

/**
 * Recursively resolve dependencies for a given version.
 * @param instanceName - Instance to check against installed mods
 * @param version - The ModrinthVersion to resolve deps for
 * @param visited - Set of already-visited project IDs (cycle detection)
 * @param depth - Current recursion depth
 */
export async function resolveDependencies(
  instanceName: string,
  version: ModrinthVersion,
  visited: Set<string> = new Set(),
  depth: number = 0,
): Promise<ResolveResult> {
  const allDeps: DependencyInfo[] = [];
  let circularDetected = false;

  // Filter to only relevant deps (skip embedded)
  const relevantDeps = version.dependencies.filter(
    (d) => d.project_id && d.dependency_type !== "embedded",
  );

  if (relevantDeps.length === 0) {
    return { dependencies: [], circularDetected: false, hasMissing: false, hasConflicts: false };
  }

  // Resolve project names AND slugs (slugs needed for Rust matching)
  const projectIds = [...new Set(relevantDeps.map((d) => d.project_id as string))];
  const metaResults = await Promise.all(projectIds.map((id) => resolveProjectMeta(id)));

  const slugMap = new Map<string, string | null>();
  const metaMap = new Map<string, { name: string; iconUrl: string | null }>();
  for (let i = 0; i < projectIds.length; i++) {
    const r = metaResults[i];
    if (r) {
      slugMap.set(projectIds[i]!, r.slug);
      metaMap.set(projectIds[i]!, { name: r.name, iconUrl: r.iconUrl });
    }
  }

  // Check installed — send SLUGS to Rust (matches against project_slug in registry)
  const statusMap = await checkInstalledStatus(
    instanceName,
    relevantDeps.map((d) => {
      const hash = d.project_id!;
      const slug = slugMap.get(hash);
      return { projectId: slug ?? hash, type: d.dependency_type };
    }),
  );

  // Process each dependency
  for (const dep of relevantDeps) {
    const pid = dep.project_id!;
    const slug = slugMap.get(pid);
    const status = statusMap.get(slug ?? pid) ?? statusMap.get(pid);
    const meta = metaMap.get(pid);

    // Cycle detection
    if (visited.has(pid)) {
      circularDetected = true;
      allDeps.push({
        projectId: pid,
        type: dep.dependency_type as DependencyInfo["type"],
        installed: status?.installed ?? false,
        installedName: status?.installedName,
        modName: meta?.name ?? pid,
        iconUrl: meta?.iconUrl ?? null,
        depth,
        children: [],
      });
      continue;
    }

    visited.add(pid);

    const info: DependencyInfo = {
      projectId: pid,
      type: dep.dependency_type as DependencyInfo["type"],
      installed: status?.installed ?? false,
      installedName: status?.installedName,
      modName: meta?.name ?? pid,
      iconUrl: meta?.iconUrl ?? null,
      depth,
    };

    // If the dep is required and not installed, recursively resolve its transitive deps
    if (dep.dependency_type === "required" && !status?.installed && depth < 3) {
      try {
        const versions = await getProjectVersions(pid, { loaders: ["fabric"] });
        if (versions.length > 0) {
          const latest = versions.find((v) => v.version_type === "release") ?? versions[0]!;
          const transitive = await resolveDependencies(instanceName, latest, visited, depth + 1);
          info.children = transitive.dependencies;
          if (transitive.circularDetected) circularDetected = true;
        }
      } catch {
        // Skip transitive resolution on error
      }
    }

    allDeps.push(info);
  }

  const hasMissing = allDeps.some(
    (d) => d.type === "required" && !d.installed,
  );
  const hasConflicts = allDeps.some(
    (d) => d.type === "incompatible" && d.installed,
  );

  return {
    dependencies: allDeps,
    circularDetected,
    hasMissing,
    hasConflicts,
  };
}

File: src-tauri/src/dependency_resolver.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

use crate::sanitize::sanitize_name;

// ─── Types ──────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DependencyStatus {
    /// The project ID of the dependency on Modrinth
    pub project_id: String,
    /// Dependency type: required, optional, incompatible, embedded
    pub dependency_type: String,
    /// Optional version ID constraint
    pub version_id: Option<String>,
    /// Whether this dependency is installed in the instance
    pub installed: bool,
    /// The installed mod name (if installed)
    pub installed_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveDependenciesResult {
    /// All dependency entries (including those already installed)
    pub dependencies: Vec<DependencyStatus>,
    /// Whether a circular dependency was detected
    pub circular_detected: bool,
    /// List of projects visited during resolution (for cycle detection)
    pub visited: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct InstalledMod {
    pub name: String,
    pub project_slug: Option<String>,
    pub file_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ModsRegistry {
    mods: Vec<InstalledMod>,
}

// ─── Helpers ────────────────────────────────────────────────────────

fn get_instance_mods_dir(app_data_dir: &std::path::Path, instance_name: &str) -> PathBuf {
    let instances_dir = app_data_dir.join("instances");
    instances_dir
        .join(sanitize_name(instance_name))
        .join("mods")
}

fn read_registry(mods_dir: &std::path::Path) -> ModsRegistry {
    let path = mods_dir.join("mods.json");
    if path.exists() {
        fs::read_to_string(&path)
            .ok()
            .and_then(|s| serde_json::from_str(&s).ok())
            .unwrap_or(ModsRegistry { mods: Vec::new() })
    } else {
        ModsRegistry { mods: Vec::new() }
    }
}

/// Check if a project_id is already installed in the instance.
/// Returns (installed, installed_name) tuple.
fn check_installed(registry: &ModsRegistry, project_id: &str) -> (bool, Option<String>) {
    for mod_entry in &registry.mods {
        if let Some(ref slug) = mod_entry.project_slug {
            if slug == project_id {
                return (true, Some(mod_entry.name.clone()));
            }
        }
    }
    (false, None)
}

/// Scan the filesystem for orphan .jar files that might be dependencies
/// without registry entries, and check if any match a project_id by file name.
fn check_filesystem(mods_dir: &std::path::Path, project_id: &str) -> bool {
    if !mods_dir.exists() {
        return false;
    }

    if let Ok(entries) = fs::read_dir(mods_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    // Check if the file name contains the project slug
                    if name.to_lowercase().contains(&project_id.to_lowercase())
                        && (name.ends_with(".jar") || name.ends_with(".jar.disabled"))
                    {
                        return true;
                    }
                }
            }
        }
    }
    false
}

// ─── Public API ─────────────────────────────────────────────────────

/// Resolve dependencies for a given list of dependency entries.
/// Checks each dependency against the installed mods in the instance.
/// 
/// This does NOT recursively resolve transitive dependencies — that's done
/// on the frontend using the Modrinth API. This command only checks which
/// of the given project IDs are already installed on disk.
pub fn resolve_dependencies(
    app_data_dir: &std::path::Path,
    instance_name: &str,
    dependencies: Vec<serde_json::Value>,
) -> Result<ResolveDependenciesResult, String> {
    let mods_dir = get_instance_mods_dir(app_data_dir, instance_name);
    let registry = read_registry(&mods_dir);

    let mut results = Vec::new();
    let mut visited = Vec::new();

    for dep in &dependencies {
        let project_id = dep.get("projectId")
            .or_else(|| dep.get("project_id"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        if project_id.is_empty() {
            continue;
        }

        let dependency_type = dep.get("dependencyType")
            .or_else(|| dep.get("dependency_type"))
            .and_then(|v| v.as_str())
            .unwrap_or("required")
            .to_string();

        let version_id = dep.get("versionId")
            .or_else(|| dep.get("version_id"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        // Check for circular dependency
        if visited.contains(&project_id) {
            return Ok(ResolveDependenciesResult {
                dependencies: results,
                circular_detected: true,
                visited,
            });
        }
        visited.push(project_id.clone());

        // Check registry first, then filesystem
        let (installed, installed_name) = check_installed(&registry, &project_id);
        let installed = installed || check_filesystem(&mods_dir, &project_id);

        results.push(DependencyStatus {
            project_id,
            dependency_type,
            version_id,
            installed,
            installed_name,
        });
    }

    Ok(ResolveDependenciesResult {
        dependencies: results,
        circular_detected: false,
        visited,
    })
}
